### ISSUE TITLE
Can't read Neuropixels NHP (commercial release) metadata

### ISSUE THREAD

**Issue by EricKenjiLee** [m-beau/NeuroPyxels#385]
Seems to be the exact problem as #378 but now with, `WARNING probe version 1030.0 not handled`. AP bin .meta file is attached with .txt extension (since GitHub doesn't allow a .meta extension; it's saved as ASCII so I would try to read this file directly or it will throw a codec error). [Tiberius_TF_PMD_0425_bank0_1_g0_t0.imec0.ap.meta.txt](https://github.com/m-beau/NeuroPyxels/files/15255959/Tiberius_TF_PMD_0425_bank0_1_g0_t0.imec0.ap.meta.txt)

SpikeGLX ver.: 20230411, Imex Phase30, v.3.60
Probe Type: NP 1032

Full traceback is as follows,

> UnboundLocalError                         Traceback (most recent call last)
> Cell In[15], line 3
>       1 ap_bin_dir = "/Volumes/Kenji3/Tiberius_TF_PMD_0425_bank0_1_g0/Tiberius_TF_PMD_0425_bank0_1_g0_imec0/"
> ----> 3 meta = npyx.read_metadata(ap_bin_dir)
> 
> File ~/Library/Caches/pypoetry/virtualenvs/cfd-dlpfc-z1Y6YtZg-py3.9/lib/python3.9/site-packages/npyx/inout.py:48, in read_metadata(dp)
>      46         meta[probe] = metadata(dpx)
>      47 else:
> ---> 48     meta = metadata(dp)
>      50 return meta
> 
> File ~/Library/Caches/pypoetry/virtualenvs/cfd-dlpfc-z1Y6YtZg-py3.9/lib/python3.9/site-packages/npyx/inout.py:244, in metadata(dp)
>     242     bits_encoding=14
>     243     ampFactor=80 # hardcoded
> --> 244 meta['bit_uV_conv_factor']=(Vrange/2**bits_encoding/ampFactor)
>     247 # find everything else
>     248 for filt_key in ['highpass','lowpass']:
> 
> UnboundLocalError: local variable 'bits_encoding' referenced before assignment

Thank you for making such a useful package!
Kenji

**Comment by m-beau** [comment]
Hi there!

Whoops sorry about that - pushing a fix asap

**Comment by m-beau** [comment]
Can you print here the contents of the .ap.meta file of this recording?

**Comment by m-beau** [comment]
Is your probe a UHD probe, or NXT, or opto-neuropixels?

**Comment by EricKenjiLee** [comment]
Actually none of the above? It's an NHP Long; the 4.5 cm primate probe.

**Comment by m-beau** [comment]
Oh right, that's the title of your issue - sorry about that, I am not familiar with primate probes.

Just to make npyx future-proof, could you share with me the list of Neuropixels probe types used in primates, and their respective SpikeGLX metadata codes (which you will find at the imDatPrb_type line in your .meta file)? Cheers!

**Comment by EricKenjiLee** [comment]
We use both the original unsharpened Neuropixels NHP (1.0 cm) and NHP Long (4.5 cm) which are type NP1010 and NP1032 respectively. A few other labs probably still have non-commercial versions of the NHP probes hanging around but I have no idea of the types.

Thanks again!
Kenji

**Comment by m-beau** [comment]
All good! You should be able to pip-install the new version (4.0.6) now.

**Comment by EricKenjiLee** [comment]
Hmm now it yields the following error when I run the following code,

```
ap_bin_dir = "/Volumes/Kenji3/Tiberius_TF_PMD_0425_bank0_1_g0/Tiberius_TF_PMD_0425_bank0_1_g0_imec0/"
meta = npyx.read_metadata(ap_bin_dir)
```

> KeyError                                  Traceback (most recent call last)
> Cell In[29], line 2
>       1 ap_bin_dir = "/Volumes/Kenji3/Tiberius_TF_PMD_0425_bank0_1_g0/Tiberius_TF_PMD_0425_bank0_1_g0_imec0/"
> ----> 2 meta = npyx.read_metadata(ap_bin_dir)
>       3 # units = npyx.gl.get_units(ap_bin_dir,quality='good')
> 
> File ~/Library/Caches/pypoetry/virtualenvs/cfd-dlpfc-z1Y6YtZg-py3.9/lib/python3.9/site-packages/npyx/inout.py:48, in read_metadata(dp)
>      46         meta[probe] = metadata(dpx)
>      47 else:
> ---> 48     meta = metadata(dp)
>      50 return meta
> 
> File ~/Library/Caches/pypoetry/virtualenvs/cfd-dlpfc-z1Y6YtZg-py3.9/lib/python3.9/site-packages/npyx/inout.py:224, in metadata(dp)
>     222 if glx_probe_version in probe_versions['glx']:
>     223     meta['probe_version'] = probe_versions['glx'][glx_probe_version]
> --> 224     meta['probe_version_int'] = probe_versions['int'][meta['probe_version']]
>     225 else:
>     226     print(f'WARNING probe version {glx_probe_version} not handled - post an issue at www.github.com/m-beau/NeuroPyxels and provide your .ap.meta file.')
> 
> KeyError: 'NHP_long_primate'

**Comment by m-beau** [comment]
Sorry about this - can you try installing from the master branch with pip install git+https://github.com/m-beau/NeuroPyxels.git ?

**Comment by m-beau** [comment]
Please let me know if it solved your issue, so that I can close this!

**Comment by EricKenjiLee** [comment]
Hi Maxime, 

Thanks for following up! I now get the following error; I think the primate probe key needs to be added here as well. Also, I think presumably the probe type 1030 is also true of the NHP shorts and may also be for the to-be-released NHP mediums (2.5 cm). Not 100% sure tho! Just thought I'd let you know in case you want to change the key from "NHP_long_primate" to something like "NHP_1.0". 

Thanks again for all your support on this tool!
Kenji

```
ap_bin_dir = "/Volumes/Kenji3/Tiberius_TF_PMD_0425_bank0_1_g0/Tiberius_TF_PMD_0425_bank0_1_g0_imec0/"
meta = npyx.read_metadata(ap_bin_dir)
```

> ValueError                                Traceback (most recent call last)
> Cell In[8], line 2
>       1 ap_bin_dir = "/Volumes/Kenji3/Tiberius_TF_PMD_0425_bank0_1_g0/Tiberius_TF_PMD_0425_bank0_1_g0_imec0/"
> ----> 2 meta = npyx.read_metadata(ap_bin_dir)
>       3 # units = npyx.gl.get_units(ap_bin_dir,quality='good')
> 
> File ~/Library/Caches/pypoetry/virtualenvs/cfd-dlpfc-z1Y6YtZg-py3.9/lib/python3.9/site-packages/npyx/inout.py:48, in read_metadata(dp)
>      46         meta[probe] = metadata(dpx)
>      47 else:
> ---> 48     meta = metadata(dp)
>      50 return meta
> 
> File ~/Library/Caches/pypoetry/virtualenvs/cfd-dlpfc-z1Y6YtZg-py3.9/lib/python3.9/site-packages/npyx/inout.py:249, in metadata(dp)
>     247     ampFactor=80 # hardcoded
>     248 else:
> --> 249     raise ValueError(f"Probe version unhandled - bits_encoding unknown.")
>     250 meta['bit_uV_conv_factor']=(Vrange/2**bits_encoding/ampFactor)
>     253 # find everything else
> 
> ValueError: Probe version unhandled - bits_encoding unknown.

**Comment by m-beau** [comment]
Right, sorry about that - it is harder to not let bugs slip through without a dataset to directly run tests!

I followed your advice, this probe's version is now coined "NHP_1.0".

The number of bits for encoding is 10, like in 1.0s, correct? I assumed so. Please install from the master branch again, and let me know here if everything works as expected.
